File: src/main/kotlin/com/lwyr/ai/service/PdfService.kt
package com.lwyr.ai.service

import dev.langchain4j.data.document.Document
import dev.langchain4j.data.document.DocumentSplitter
import dev.langchain4j.data.document.splitter.DocumentByCharacterSplitter
import dev.langchain4j.data.segment.TextSegment
import mu.KotlinLogging
import org.apache.pdfbox.Loader
import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.text.PDFTextStripper
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.security.MessageDigest

private val logger = KotlinLogging.logger(PdfService::class.java.name)

@Service
class PdfService(
    @Value("\${pdf.chunk-size:1000}")
    private val chunkSize: Int,

    @Value("\${pdf.chunk-overlap:200}")
    private val chunkOverlap: Int
) {

    private val documentSplitter: DocumentSplitter = DocumentByCharacterSplitter(
        chunkSize,
        chunkOverlap
    )

    fun extractText(inputStream: InputStream, filename: String): String {
        logger.info { "Extracting text from PDF: $filename" }
        val bytes = inputStream.readBytes()
        val pdfDocument: PDDocument = Loader.loadPDF(bytes)
        val text = pdfDocument.use { doc ->
            PDFTextStripper().getText(doc)
        }
        logger.info { "Extracted ${text.length} characters from PDF" }
        return text
    }

    fun chunkText(text: String, filename: String): List<TextSegment> {
        logger.info { "Chunking text from $filename into segments" }
        val document = Document.from(text)
        val segments = documentSplitter.split(document)
        logger.info { "Created ${segments.size} chunks from $filename" }
        return segments
    }

    fun calculateChecksum(inputStream: InputStream): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val buffer = ByteArray(8192)
        var bytesRead: Int
        while (inputStream.read(buffer).also { bytesRead = it } != -1) {
            digest.update(buffer, 0, bytesRead)
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }
}

File: app/services/pdf_service.py
"""PDF processing service."""

import hashlib
import logging
from io import BytesIO

from pypdf import PdfReader

from app.config import get_settings

logger = logging.getLogger(__name__)
settings = get_settings()


class PdfService:
    """Service for PDF processing operations."""

    def __init__(
        self,
        chunk_size: int = settings.pdf_chunk_size,
        chunk_overlap: int = settings.pdf_chunk_overlap,
    ) -> None:
        self.chunk_size = chunk_size
        self.chunk_overlap = chunk_overlap

    def extract_text(self, file_content: bytes, filename: str) -> str:
        """Extract text from a PDF file."""
        logger.info(f"Extracting text from PDF: {filename}")
        pdf_document = PdfReader(BytesIO(file_content))
        text_parts: list[str] = []

        for page in pdf_document.pages:
            page_text = page.extract_text()
            if page_text:
                text_parts.append(page_text)

        text = "\n".join(text_parts)
        logger.info(f"Extracted {len(text)} characters from PDF: {filename}")
        return text

    def chunk_text(self, text: str, filename: str) -> list[str]:
        """Split text into overlapping chunks."""
        logger.info(f"Chunking text from {filename}")
        chunks: list[str] = []
        start = 0

        while start < len(text):
            end = start + self.chunk_size
            chunk = text[start:end]
            chunks.append(chunk)

            if end >= len(text):
                break

            start = end - self.chunk_overlap

        logger.info(f"Created {len(chunks)} chunks from {filename}")
        return chunks

    def calculate_checksum(self, file_content: bytes) -> str:
        """Calculate SHA-256 checksum of file content."""
        digest = hashlib.sha256()
        digest.update(file_content)
        return digest.hexdigest()
